### PR TITLE
Encrypt secret key from KeyFile off staging

### PR DESCRIPTION
**Pull request by cyberbono3**
I submit my drafr PR for your review.
Feedback required.

perform_encryption performs the following operations:
if there are no home folder:
- it asks to enter password to encrypt the private key with
- saves onto the disk an encrypted private key
if there is home folder already:
- reads the encrypted private key
- asks user for the password
- decrypts the private key
- uses it to sign the blocks (TODO)

### CODE REVIEW

**Comment by ilblackdragon** [comment]
Let's rebase with staging (move this into the crypto crate keyfile.rs).

**Comment by ilblackdragon** [comment]
Unclear why so many changes o_O
I just merge yours and did git diff HEAD~1 to actually see your changes.

PasswordInput::enter_from_console -> should support that user might not want to encrypt the key.
`MAX_LENGTH` is a constant, why do you need it to be also a field?

to pad you can use use:
```
let buf = [0u8; 32];
buf.copy_from_slice(self.content[..]);
buf.to_vec()
```

should we just add support for KeyFile to `write_to_encrypted_file`?

Also all this needs tests / actual usage - it's hard to judge about APIs when there is no usage. E.g. it feel like encrypt_keyfile_secretkey / processed_serialized_file will be awkward to use.

Ideally let's make KeyFile always encrypted and just support "empty" password as a special case of non encrypted version. And then have interactive APIs (from_file_with_password() that asks for password if needed) and non-interactive APIs (from_file(path, password)).

**Comment by cyberbono3** [comment]
>And then have interactive APIs (from_file_with_password() that asks for password if needed) and non-interactive APIs (from_file(path, password)).

can you give me more information about functionality of **from_file_with_password()** and **from_file(path, password)**  ?

**Comment by ilblackdragon** [comment]
`from_file(path, password)` reads the file and the decodes it with the password.
`from_file_with_password(path)` asks for password in command line and then decodes with that password.

**Comment by cyberbono3** [comment]
have some progress, feedback required

TODO: 
-  check if decryption password matches the one used for encryption
-  implement methods and traits SecretKeyView and  PublicKeyView
-  testing

Questions:

1. what is best way to check if decryption password matches the one used for encryption?
2. what methods and traits should be implemented for SecretKeyView and PublicKeyView?

**Comment by k06a** [comment]
Please revert the following files:
```
chain/client/src/lib.rs
chain/jsonrpc/src/lib.rs
chain/network/src/lib.rs
chain/network/src/rate_counter.rs
core/store/benches/trie_bench.rs
near/src/main.rs
```

File: core/primitives/src/crypto/signer.rs [hunk 1]
```diff
@@ -53,15 +62,209 @@ impl KeyFile {
             panic!("Failed to write a key file {}", err);
         }
     }
+    // is_encrypted checks if current keyhash struct matches default one
```
Comment (mfornet) [inline core/primitives/src/crypto/signer.rs]: Doc comments should use three / in rust.

Consider using to rustfmt for code formatting.

**Comment by mfornet** [comment]
@cyberbono3 to detect if a secret key was indeed used to encrypt the data you can sign a hash of the data (or whatever other hash) and then verify the signature. 
Similarly, you can encrypt a small known message and check it matches given secret key.

Both of this alternatives allow to do very fast brute force, since it allows to figure the password from small data, so the key length/algorithm used needs to be strong enough.

**Comment by ilblackdragon** [comment]
Closing this, as #1336 is the new version

**Pull request closed without merging**